### PR TITLE
#166272340 user can create, modify, delete and read one and read all article

### PR DESCRIPTION
**Pull request by emmygozi**
#### What does this PR do?

User can create an article with images if they choose
#### Description of Task to be completed?

- user should be able to create an article
- user should be able to update an article
- user should be able to get one article
- user should be able to get all available articles
- user should be able to delete an article
- an article should have a unique slug
- an article should have a timestamp
- created migration to add a column for soft delete
- an article can have images
- an article should return the author details
- an article should return user ratings
- write a unit test for the created user and test all edge cases
- write a unit test for the updated user and test all edge cases
- write a unit test for the deleted user and test all edge cases
- write a unit test to get one user
- write a unit test to get all user
- refactor findOne and findAll articles method to a different file
- Add relation to get an article created by a user using a magic function `hasArticle`

#### How should this be manually tested?

Navigate to the endpoint:
  `POST, GET http://localhost:3000/api/v1/articles`
    `GET PUT DELETE http://localhost:3000/api/v1/articles/:slug`
#### Any background context you want to provide?

N/A

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/166272340

#### Screenshots (if appropriate)

image

#### Questions:

N/A

### CODE REVIEW

**Review by despeauxz — DISMISSED** [review]
Nice job, just some little comments. Please fix the hound issue, also the task specified CRUD operations for the article and need to document the endpoint on swagger. Generally, nice job

File: server/app.js [hunk 1]
```diff
@@ -21,7 +22,7 @@ app.use(bodyParser.urlencoded({
 }));
 
 app.use('/api/v1', apis);
-
+app.use('/api/v1', articlesRoutes);
```
Comment (fidelisojeah) [inline server/app.js]: what is this doing here exactly?

File: server/models/article.js [hunk 2]
```diff
@@ -1,3 +1,5 @@
+import slugCreator from '../helpers/slugCreator';
```
Comment (fidelisojeah) [inline server/models/article.js]: your path fix this
```suggestion
import slugCreator from '@helpers/slugCreator';
```

File: server/controllers/articles.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      const articleDetails = await validateArticle(req.body);
+      const { id: userId } = req.decoded;
+      const { title } = articleDetails;
+      const createArticleDetails = {
+        slug: title, userId, ...articleDetails
+      };
+
+      const createdArticle = await Article.create(createArticleDetails);
+      const { dataValues: article } = createdArticle;
+      return res.status(201).json({ status: 'success', message: 'Article created successfully', article });
```
Comment (fidelisojeah) [inline server/controllers/articles.js]:
```suggestion
      return res.status(201).json({ status: 'success', message: 'Article created successfully', payload: article });
```

**Review by fidelisojeah — COMMENTED** [review]
Good job

Just a few changes

File: server/app.js (hunk 1 above)
Comment (emmygozi) [reply server/app.js]: Thank you for your observation. Fixed.

File: server/controllers/articles.js [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+  * @return {json} Returns json object
+  * @static
+  */
+  static async update(req, res, next) {
+    try {
+      const articleDetails = await validateArticle(req.body);
+      const { id: userId } = req.decoded;
+      const { slug: updatedSlug } = req.params;
+
+      const user = await User.findByPk(userId);
```
Comment (fidelisojeah) [inline server/controllers/articles.js]: make use of `req.user` here please

File: server/controllers/articles.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+      const canUpdate = await user.hasArticle(getArticle);
+
+
+      if (!getArticle) {
+        return Response.error(res, 404, 'Article does not exist');
+      }
+
+      if (!canUpdate) {
+        return Response.error(res, 401, 'You do not have permission to update this article!');
```
Comment (fidelisojeah) [inline server/controllers/articles.js]: this is a `403` not `401`

File: server/controllers/articles.js [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+      const getArticle = await Article.findOne({ where: { slug } });
+
+      if (!getArticle) {
+        return Response.error(res, 404, 'Article does not exist');
+      }
+
+      const deletedArticle = await Article.destroy({ where: { slug, userId } });
+
+      if (!deletedArticle) {
+        return Response.error(res, 401, 'You do not have permission to delete this article!');
```
Comment (fidelisojeah) [inline server/controllers/articles.js]: same here `403`

File: server/helpers/articlePayload.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import sequelize from 'sequelize';
+import models from '@models';
+
+const { User, Rating } = models;
+
+
+const articlePayload = {
```
Comment (fidelisojeah) [inline server/helpers/articlePayload.js]: get the updated version of this payload from @EMacco

**Review by fidelisojeah — DISMISSED** [review]
Good job, just a few changes to make

File: server/controllers/articles.js (hunk 4 above)
Comment (emmygozi) [reply server/controllers/articles.js]: Would fix that now.

File: server/helpers/articlePayload.js (hunk 7 above)
Comment (emmygozi) [reply server/helpers/articlePayload.js]: Syncing with @EMacco . Thanks for your kind observation.

**Review by despeauxz — APPROVED** [review]

**Review by EMacco — APPROVED** [review]

**Review by fidelisojeah — APPROVED** [review]
